lib/runtime/version.go

```diff
@@ -4,89 +4,32 @@
 package runtime
 
 import (
+	"errors"
+	"fmt"
+	"strings"
+
 	"github.com/ChainSafe/gossamer/pkg/scale"
 )
 
-//go:generate mockery --name Version --structname Version --case underscore --keeptree
-
-// Version represents the data returned by runtime call core_version
-type Version interface {
-	SpecName() []byte
-	ImplName() []byte
-	AuthoringVersion() uint32
-	SpecVersion() uint32
-	ImplVersion() uint32
-	APIItems() []APIItem
-	TransactionVersion() uint32
-	Encode() ([]byte, error)
-}
-
 // APIItem struct to hold runtime API Name and Version
 type APIItem struct {
 	Name [8]byte
 	Ver  uint32
 }
 
-// LegacyVersionData is the runtime version info returned by legacy runtimes
-type LegacyVersionData struct {
-	specName         []byte
-	implName         []byte
-	authoringVersion uint32
-	specVersion      uint32
-	implVersion      uint32
-	apiItems         []APIItem
-}
-
-// NewLegacyVersionData returns a new LegacyVersionData
-func NewLegacyVersionData(specName, implName []byte,
-	authoringVersion, specVersion, implVersion uint32,
-	apiItems []APIItem) *LegacyVersionData {
-	return &LegacyVersionData{
-		specName:         specName,
-		implName:         implName,
-		authoringVersion: authoringVersion,
-		specVersion:      specVersion,
-		implVersion:      implVersion,
-		apiItems:         apiItems,
-	}
-}
-
-// SpecName returns the spec name
-func (lvd *LegacyVersionData) SpecName() []byte {
-	return lvd.specName
-}
-
-// ImplName returns the implementation name
-func (lvd *LegacyVersionData) ImplName() []byte {
-	return lvd.implName
-}
-
-// AuthoringVersion returns the authoring version
-func (lvd *LegacyVersionData) AuthoringVersion() uint32 {
-	return lvd.authoringVersion
-}
-
-// SpecVersion returns the spec version
-func (lvd *LegacyVersionData) SpecVersion() uint32 {
-	return lvd.specVersion
-}
-
-// ImplVersion returns the implementation version
-func (lvd *LegacyVersionData) ImplVersion() uint32 {
-	return lvd.implVersion
-}
-
-// APIItems returns the API items
-func (lvd *LegacyVersionData) APIItems() []APIItem {
-	return lvd.apiItems
-}
-
-// TransactionVersion returns the transaction version
-func (*LegacyVersionData) TransactionVersion() uint32 {
-	return 0
+// Version is the runtime version info.
+type Version struct {
+	SpecName           []byte
+	ImplName           []byte
+	AuthoringVersion   uint32
+	SpecVersion        uint32
+	ImplVersion        uint32
+	APIItems           []APIItem
+	TransactionVersion uint32
+	legacy             bool
 }
 
-type legacyVersionData struct {
+type legacyData struct {
```

Comment:
why not embed this struct in `Version`?

Reply:
I tried it, it encodes wrong with scale, so we would need another clone struct. But good suggestion nonetheless!!

Reply:
Now irrelevant, this is changed to be:

- encoding: always encode to last format with all fields (some to their zero value `0`)
- decoding: decodes all required fields and tries to decode optional fields one by one without erroring when reaching EOF